### DISCUSSION TITLE
New ESP - Azure Communication Services

### DISCUSSION THREAD

**Discussion by f-reinsch** [anymail/django-anymail#408] [Ideas]
Hi,

are there any ongoing plans to integrate Azure Communication Services as new ESP?
Would be a good alternative for Amazon SES from a pricing perspective 😄

**Comment by medmunds** [comment]
I'm not aware of anyone working on an Azure Communication Services integration, but it looks like a reasonable fit for Anymail. Others can express support by upvoting the first message in this thread.

Configuring Azure Communication Email seems at least as complicated as configuring Amazon SES. Like Amazon, Microsoft seems to have a preference for inventing their own terminology. (E.g., "[P2 sender email address](https://learn.microsoft.com/en-us/azure/communication-services/concepts/email/prepare-email-communication-resource#:~:text=Define%20the%20P2%20sender%20email%20address%20that%20most%20email%20recipients%20see%20on%20their%20email%20client.)" for what the rest of us would call a "From header")

I'm unlikely to get to this in the next few months; if you (or anyone else) is interested in working on it, let me know. In the meantime, it should be possible to send from Django through Azure by using [Azure's SMTP option](https://learn.microsoft.com/en-us/azure/communication-services/concepts/email/email-smtp-overview) and Django's built-in SMTP EmailBackend.

More Azure info & links:
- Email sending
  - [HTTP API](https://learn.microsoft.com/en-us/rest/api/communication/dataplane/email/send?view=rest-communication-dataplane-2023-03-31&tabs=HTTP) - requires a [signed request](https://learn.microsoft.com/en-us/rest/api/communication/authentication?view=rest-communication-dataplane-2023-03-31#signing-an-http-request)
  - [Python SDK](https://learn.microsoft.com/en-us/python/api/overview/azure/communication-email-readme?view=azure-python) - handles request signing, but appears to require polling? ([tutorial](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/email/send-email?tabs=windows%2Cconnection-string%2Csend-email-and-get-status-async%2Csync-client&pivots=programming-language-python))
  - Does not seem to support inline images, tags, metadata, templates, batch sending
- Email tracking: [email events](https://learn.microsoft.com/en-us/azure/event-grid/communication-services-email-events), delivered to Azure Event Grid, which can notify a webhook
- Inbound email: not supported? (unless Azure has named it something non-obvious)

**Reply by f-reinsch** [reply]
Hi @medmunds, 

thank you for the fast and detailed reply, awesome!

Unfortunatly i´m currently also not able to invest heavily in this, you know end of the year stress and so forth :-)
